### PR TITLE
TEAL: assembler error on branching behind last instruction

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

* TEAL v1 does not allow such program, so make assembler compatible
* TEAL v2 permits the mentioned behavior

## Test Plan

Added new test to validate v1 and v2 behavior

## Notes

1. Fixed disassembly listing for such programs - no unknown labels anymore
2. Switched some asm tests to `t.Parallel()`

### CODE REVIEW

**Comment by algorandskiy** [comment]
Fix for #1212

File: data/transactions/logic/assembler.go [hunk 1]
```diff
@@ -1298,6 +1308,17 @@ func (dis *disassembleState) putLabel(label string, target int) {
 	dis.pendingLabels[target] = label
 }
 
+func (dis *disassembleState) outputLabelIfNeeded() error {
```
Comment (tsachiherman) [inline data/transactions/logic/assembler.go:1311]: nit: coding style - 
```go
func (dis *disassembleState) outputLabelIfNeeded() (err error) {
 	if label, hasLabel := dis.pendingLabels[dis.pc]; hasLabel {
 		_, err := fmt.Fprintf(dis.out, "%s:\n", label)
 	}
 	return
 }
```

File: data/transactions/logic/assembler.go [hunk 2]
```diff
@@ -1072,6 +1072,16 @@ func (ops *OpStream) assemble(fin io.Reader) error {
 		err := fmt.Errorf("unknown opcode %v", opstring)
 		return lineErr(ops.sourceLine, err)
 	}
+
+	// backward compatibility: do not allow jumps behind last instruction in TEAL v1
+	if ops.Version <= AssemblerDefaultVersion {
```
Comment (brianolson) [inline data/transactions/logic/assembler.go:1077]: Shouldn't this simply be `<= 1`

Comment (brianolson) [reply data/transactions/logic/assembler.go:1077]: `AssemblerDefaultVersion` will change, but we want to pin this at <=1

Comment (algorandskiy) [reply data/transactions/logic/assembler.go]: done

File: data/transactions/logic/assembler.go (hunk 1 above)
Comment (algorandskiy) [reply data/transactions/logic/assembler.go]: done

**Comment by algorandskiy** [comment]
Fixed `TestBackwardCompatAssemble` a bit to extend the test beyond v2

**Pull request merged**